File: 0295-find-median-from-data-stream/solution.py
class MedianFinder:

    def __init__(self):
        self.minheap = []
        self.maxheap = []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.minheap, num * -1)
        if self.minheap and self.maxheap:
            if (self.minheap[0] * -1) > (self.maxheap[0]):
                val = -1 * heapq.heappop(self.minheap)
                heapq.heappush(self.maxheap, val)
            elif (len(self.minheap) - len(self.maxheap)) > 1:
                val = -1 * heapq.heappop(self.minheap)
                heapq.heappush(self.maxheap, val)
        if (len(self.maxheap) - len(self.minheap)) > 1:
            val = -1 * heapq.heappop(self.maxheap)
            heapq.heappush(self.minheap, val)
        elif (len(self.minheap) - len(self.maxheap)) > 1:
            val = -1 * heapq.heappop(self.minheap)
            heapq.heappush(self.maxheap, val)

    def findMedian(self) -> float:
        if len(self.minheap) > len(self.maxheap):
            return self.minheap[0] * -1
        elif len(self.maxheap) > len(self.minheap):
            return self.maxheap[0]
        else:
            return ((self.minheap[0]  * -1) + self.maxheap[0])/2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 0295-find-median-from-data-stream/solution.cpp
class MedianFinder {
private:
    priority_queue<int> maxq; //left side
    priority_queue<int, vector<int>, greater<int>> minq; // right side
public:
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        maxq.push(num);
        minq.push(maxq.top());
        maxq.pop();

        if(minq.size() > maxq.size()){
            maxq.push(minq.top());
            minq.pop();
        }
    }
    
    double findMedian() {
        if(minq.size() > maxq.size()){
            return minq.top();
        }
        else if(maxq.size() > minq.size()){
            return maxq.top();
        }
        double res = (minq.top() + maxq.top()) / 2.0;
        return res;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
